storage: avoid errant Raft snapshots during splits

tbg opened the pull request:
A known race occurs during splits when some nodes apply the split
trigger faster than others. The "slow" node(s) may learn about the
newly created right hand side replica through Raft messages arriving
from the "fast" nodes. In such cases, the leader will immediately try to
catch the follower (which it sees at log position zero) up via a
snapshot, but this isn't possible since there's an overlapping replica
(the pre-split replica waiting to apply the trigger). I *think* we
would find that out early in the life of the snapshot, i.e. we wouldn't
transfer large amounts of data around. Nevertheless, this smells bad.

The following script reproduces this by running 1000 splits back
to back in a three node local cluster. In an ideal world, we'd
see zero Raft snapshots here, but we see north of a hundred typically.

Not all of those are due to the race. ReplicaGC and Raft log truncation
can also contribute to this issue (disable both of those queues for pure
results).

We shouldn't ask for a snapshot when we suspect the target replica is
going to catch up on a split trigger. This commit doesn't really fix
that, but it suggests that one cheap way out is to simply report the
requested snapshot as completed immediately. Raft will go back to
probing the commit index, and this will continue until the split
trigger has won the race (of course this trick can only ever help
in situations in which we know in our heart of hearts that there's
a commit trigger that will render the snapshot useless).

```

set -euxo pipefail

killall -9 cockroach || true
killall -9 workload || true
sleep 1
rm -rf cockroach-data || true
mkdir -p cockroach-data

./cockroach start --insecure --host=localhost --port=26257 --http-port=26258 --store=cockroach-data/1 --cache=256MiB --background
./cockroach start --insecure --host=localhost --port=26259 --http-port=26260 --store=cockroach-data/2 --cache=256MiB --join=localhost:26257 --background
./cockroach start --insecure --host=localhost --port=26261 --http-port=26262 --store=cockroach-data/3 --cache=256MiB --join=localhost:26257 --background

sleep 5

./cockroach sql --insecure -e 'alter range default configure zone using range_max_bytes = 1048471142400;'
./cockroach sql --insecure -e 'set cluster setting kv.range_merge.queue_enabled = false;'
./bin/workload run kv --splits 1000 --init --drop --max-ops 1

for port in 26257 26259 26261; do
  ./cockroach sql --insecure -e "select name, value from crdb_internal.node_metrics where name like '%raftsn%' order by name desc" --port "${port}"
done

sleep 5
tail -Fq cockroach-data/*/logs/cockroach.log
```

Touches #32046 
Touches #31409 

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/31875)
<!-- Reviewable:end -->

tschottdorf commented:
This happens pretty regularly in roachtest restores too. We'll see lots of snapshots at applied index 11, often blocked by the unsplit RHS on the receiving node, which then gets added to the replica GC queue. It seems mostly harmless, but we should alleviate the problem.

tschottdorf commented:
I added a second commit that introduces heuristics that *almost* removes all of the snapshots and isn't obviously going to prevent real snapshots from being sent.

In essence, I'm trying to figure out whether the range was recently split, and whether the follower is likely to get a "free snapshot" via the split trigger very soon.

This works almost 100% of the time, except in the case in which the LHS bumps its generation rapidly (due to being split a second time). This is entering the area of diminishing returns, but I do want to get rid of these snapshots reliably when they're not essential.

This raises the question: if not via the generation counter, what's another indicator that the range has "very recently" split? Sure, I could go and read the history of the local range descriptor and see if the first (generation-zero) value is "recent". But maybe there's something less invasive I can work with.

@benesch, I suspect that if there is something nice then you'd know about it.

tschottdorf commented:
Filed https://github.com/cockroachdb/cockroach/issues/31947 about some general concerns about the Raft log queue.

tschottdorf commented:
This is still WIP, but I think worth taking a look at.

tschottdorf commented:
(in particular I haven't vetted whether it's OK to lie to Raft and just tell it it got a snapshot, though it seems to work just fine).

petermattis reviewed (dismissed):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31875)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 5119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LPrIZto13KqCuqVYWUH:-LPrIZto13KqCuqVYWUI:bms20hi) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd3a65dbafed1b9300490b38e0bbb1732d4dc264/pkg/storage/replica.go#L5119)):*
> ```Go
> 		recentlyCreatedViaSplit := desc.IsInitialized() && (desc.Generation == nil || *desc.Generation == 0) &&
> 			raftStatus != nil && raftStatus.Progress[msg.To].Match == 0 &&
> 			ticks <= 15
> ```

Newly create Raft groups have a small Raft log index, right? Can that be used here instead of looking at `Generation`?
___
*[pkg/storage/replica.go, line 5123 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LPrIHLD0jMIAw6kG0nE:-LPrIHLD0jMIAw6kG0nF:bilmdfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd3a65dbafed1b9300490b38e0bbb1732d4dc264/pkg/storage/replica.go#L5123)):*
> ```Go
> 		if recentlyCreatedViaSplit {
> 			r.withRaftGroup(false, func(rn *raft.RawNode) (bool, error) {
> 				rn.ReportSnapshot(msg.To, raft.SnapshotFinish)
> ```

`SnapshotFinish` indicates that the snapshot succeeded. That doesn't seem right.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31875)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 5123 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LPrIHLD0jMIAw6kG0nE:-LPubSZxA-V9-8mzQyL2:b3iih5j) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd3a65dbafed1b9300490b38e0bbb1732d4dc264/pkg/storage/replica.go#L5123)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`SnapshotFinish` indicates that the snapshot succeeded. That doesn't seem right.
</blockquote></details>

That's the part that we have to make "correct" (it might be correct today, but certainly doesn't look the part). If we say the snapshot failed, (I think - but the experiment suggests it) Raft is going to ask for another snapshot until it gets one. But here we have this situation where it receives a "pseudosnapshot" via the split trigger. So we lie and say "hey you got a snapshot" and apparently Raft uses this information in a way that is compatible with what we're trying to achieve (i.e. it's like hey, I seem to have gotten this follower back on track, let's try to append to him again but not using what I think would've been in the snapshot I asked for, but just using what I knew about this follower before).


<!-- Sent from Reviewable.io -->

benesch commented:
Nothing nice that I know about. We could write a range-ID local key with the time at which the range was created, but I don’t think that would be any much more precise than what you’ve got or what Peter suggested.

Here’s an alternate proposal, though, just for kicks. Whenever you split a range, you record the LHS’s range ID and applied index on the RHS. Then, whenever you send a snapshot, as part of the handshake, you send along the recorded range ID and applied index. The receiving store then checks to see if it has a replica of that range at an earlier applied index. If it does, it blocks the snapshot for a bit (before we actually do any of the hard work of generating the snap) in the hopes that it’s replica of the LHS will apply the split trigger. If the LHS replica doesnt make progress in a few hundred ms, then we queue it for GC; only if it gets GC’d or advances past the split trigger with a snapshot of its own do we allow the RHS snapshot through.

There’s some additional complexity here to avoid stalling out all four goroutines in the Raft snapshot queue, but I think it could be made to work.

tschottdorf commented:
I thought about something like that but it seems overly complex. Also, the thing I'm really after is to never have to do any work for this case in the raft snapshot queue. We should see zero Raft snapshots in all situations which don't actually require recovery.

I wonder if we can attack the problem from the other side. The snapshot gets requested by the leader after the follower rejects its MsgApp. Could we selectively drop these rejection messages for a little bit to let the split trigger "always win"? I'll find out.

tschottdorf commented:
Added a commit. It works 100% in my limited test. What I'm less sure about is that I got all of the conditions right (how do I tell whether a replica has a preemptive snapshot but isn't active yet? I think it's that it's not in its own RangeDescriptor), but I think this does look promising.

benesch reviewed (commented):
> Reviewed 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31875)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 5115 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQ-Lbzf3ylwmFtcgJGP:-LQ-LbzgBypXWFyBvZCM:btyb9yn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff5e60d70dfd3422dc474dddcbae9abf17f853c7/pkg/storage/replica.go#L5115)):*
> ```Go
> 		r.mu.RUnlock()
> 
> 		if !desc.IsInitialized() && len(desc.Replicas) == 0 && msg.RejectHint == 0 && ticks < 10 {
> ```

Ah, this feels much cleaner.

I think, though I'm not sure, that the check for `!desc.IsInitialized()` is redundant with the check for `len(desc.Replicas) == 0`. I think replicas with only a preemptive snapshot will return true for `desc.IsInitialized()`.
___
*[pkg/storage/replica.go, line 5145 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQ-PxQA1jl-0hq7YDqe:-LQ-PxQA1jl-0hq7YDqf:bls5fup) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff5e60d70dfd3422dc474dddcbae9abf17f853c7/pkg/storage/replica.go#L5145)):*
> ```Go
> 			// TODO(tschottdorf): preemptive snapshots may also be
> 			// garbage-collected before the replica addition is complete. In
> 			// such a case, we need to accept the Raft snapshot ASAP.
> ```

Can we just sidestep this race by refusing to GC replicas with preemptive snapshots for some time period after the snapshot was applied? Say, 60s?
___
*[pkg/storage/replica.go, line 5146 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQ-LliN143I2YCp0yzR:-LQ-LliN143I2YCp0yzS:b-s5vcy4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff5e60d70dfd3422dc474dddcbae9abf17f853c7/pkg/storage/replica.go#L5146)):*
> ```Go
> 			// garbage-collected before the replica addition is complete. In
> 			// such a case, we need to accept the Raft snapshot ASAP.
> 			log.Warningf(ctx, "dropping rejection from %d to %d", msg.Index, msg.RejectHint)
> ```

Not sure how WIP this line is, but a warning seems overly strong.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31875)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 5119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LPrIZto13KqCuqVYWUH:-LPudZ-1DkqgYotbiwpP:b-vxd1n0) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd3a65dbafed1b9300490b38e0bbb1732d4dc264/pkg/storage/replica.go#L5119)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Newly create Raft groups have a small Raft log index, right? Can that be used here instead of looking at `Generation`?
</blockquote></details>

Obsolete, but here's my old comment:

I had added that at some point, but the log index can race ahead pretty quickly. It wasn't clear to me where to cut it off.
___
*[pkg/storage/replica.go, line 5123 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LPrIHLD0jMIAw6kG0nE:-LQ-TaJs0nqMqv9TgWsH:bwv7l5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd3a65dbafed1b9300490b38e0bbb1732d4dc264/pkg/storage/replica.go#L5123)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

That's the part that we have to make "correct" (it might be correct today, but certainly doesn't look the part). If we say the snapshot failed, (I think - but the experiment suggests it) Raft is going to ask for another snapshot until it gets one. But here we have this situation where it receives a "pseudosnapshot" via the split trigger. So we lie and say "hey you got a snapshot" and apparently Raft uses this information in a way that is compatible with what we're trying to achieve (i.e. it's like hey, I seem to have gotten this follower back on track, let's try to append to him again but not using what I think would've been in the snapshot I asked for, but just using what I knew about this follower before).
</blockquote></details>

Thankfully this is now obsolete.
___
*[pkg/storage/replica.go, line 5115 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQ-Lbzf3ylwmFtcgJGP:-LQ-TJnICmSt9AtLyL-b:bcbhwg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff5e60d70dfd3422dc474dddcbae9abf17f853c7/pkg/storage/replica.go#L5115)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Ah, this feels much cleaner.

I think, though I'm not sure, that the check for `!desc.IsInitialized()` is redundant with the check for `len(desc.Replicas) == 0`. I think replicas with only a preemptive snapshot will return true for `desc.IsInitialized()`.
</blockquote></details>

`IsInitialized` just checks for nonempty `EndKey`, but after a preemptive snapshot, don't you have "all of the data" except you're not a member of the `RangeDescriptor`?
___
*[pkg/storage/replica.go, line 5145 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQ-PxQA1jl-0hq7YDqe:-LQ-TRHL2sek0FXmppPe:b-d5fjb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff5e60d70dfd3422dc474dddcbae9abf17f853c7/pkg/storage/replica.go#L5145)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Can we just sidestep this race by refusing to GC replicas with preemptive snapshots for some time period after the snapshot was applied? Say, 60s?
</blockquote></details>

Yes, I think so (haven't thought about the precise mitigation yet, maybe we can do something smarter, who knows). Just didn't wanna jam this into this PR. But I should land it before this one.
___
*[pkg/storage/replica.go, line 5146 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQ-LliN143I2YCp0yzR:-LQ-TYzS9oGuruLhw_jd:bivlev7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff5e60d70dfd3422dc474dddcbae9abf17f853c7/pkg/storage/replica.go#L5146)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Not sure how WIP this line is, but a warning seems overly strong.
</blockquote></details>

Definitely WIP. Thanks for pointing it out.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31875)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 5115 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQ-Lbzf3ylwmFtcgJGP:-LQ-f93q37iIJ0dZVN6x:bae54pc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff5e60d70dfd3422dc474dddcbae9abf17f853c7/pkg/storage/replica.go#L5115)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

`IsInitialized` just checks for nonempty `EndKey`, but after a preemptive snapshot, don't you have "all of the data" except you're not a member of the `RangeDescriptor`?
</blockquote></details>

Ditto that this is cleaner than trying to reject the snapshot.

I imagine that you'll be adding a test for this behavior. Writing that test isn't going to be fun, but your future self will thank you.


<!-- Sent from Reviewable.io -->

benesch reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31875)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 5115 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQ-Lbzf3ylwmFtcgJGP:-LQ-lISu2LTI6oGgArkC:b-cqpyuo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff5e60d70dfd3422dc474dddcbae9abf17f853c7/pkg/storage/replica.go#L5115)):*
> IsInitialized just checks for nonempty EndKey, but after a preemptive snapshot, don't you have "all of the data" except you're not a member of the RangeDescriptor?

Right. So if you have a preemptive snapshot, then IsInitialized is true, and so we correctly won't take this branch because we know the replica was not created as the result of a split trigger.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31875)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 5115 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQ-Lbzf3ylwmFtcgJGP:-LQ-gvuZ2C4xVFW9Ct8u:buizbhk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff5e60d70dfd3422dc474dddcbae9abf17f853c7/pkg/storage/replica.go#L5115)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

> IsInitialized just checks for nonempty EndKey, but after a preemptive snapshot, don't you have "all of the data" except you're not a member of the RangeDescriptor?

Right. So if you have a preemptive snapshot, then IsInitialized is true, and so we correctly won't take this branch because we know the replica was not created as the result of a split trigger.
</blockquote></details>

The checks here seem redundant to me too. They're all equivalent: an uninitialized replica will have empty `desc.Replicas` and a zero log index, while an initialized one is guaranteed to have non-empty `desc.Replicas` and non-zero log index. Just one of the first three checks should be sufficient (in combination with `ticks < 10`). 

If we're sending a MsgAppResp, the leader sent us a MsgApp, so the leader believes we're a member of the group. That means we're either in the "split, but not yet triggered" case or the "received preemptive snapshot and immediately GC'd it" case. I don't see a good way to distinguish the two cases at this stage, although a guard against GC'ing a brand new replica would go a long way to minimizing the risk (and `mu.ticks` gives us a good way to handle this). 


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
I need to add a test, but the non-test code is ready for review (I'm already using it in my roachtesting).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-:-LQBLqRS9Kn3pF4PEBuH:b-ak65pz)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 5115 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQ-Lbzf3ylwmFtcgJGP:-LQA2IPb71EJNayjpMVM:b4j7aua) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff5e60d70dfd3422dc474dddcbae9abf17f853c7/pkg/storage/replica.go#L5115)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The checks here seem redundant to me too. They're all equivalent: an uninitialized replica will have empty `desc.Replicas` and a zero log index, while an initialized one is guaranteed to have non-empty `desc.Replicas` and non-zero log index. Just one of the first three checks should be sufficient (in combination with `ticks < 10`). 

If we're sending a MsgAppResp, the leader sent us a MsgApp, so the leader believes we're a member of the group. That means we're either in the "split, but not yet triggered" case or the "received preemptive snapshot and immediately GC'd it" case. I don't see a good way to distinguish the two cases at this stage, although a guard against GC'ing a brand new replica would go a long way to minimizing the risk (and `mu.ticks` gives us a good way to handle this). 
</blockquote></details>

Ah, yeah, that was pretty confused. Removed.
___
*[pkg/storage/replica.go, line 5145 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQ-PxQA1jl-0hq7YDqe:-LQADLR4C9F4BDLuYfVr:brrsqfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff5e60d70dfd3422dc474dddcbae9abf17f853c7/pkg/storage/replica.go#L5145)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Yes, I think so (haven't thought about the precise mitigation yet, maybe we can do something smarter, who knows). Just didn't wanna jam this into this PR. But I should land it before this one.
</blockquote></details>

Heads up that I have a WIP for this that I'm testing with. Will send that separately and make sure it lands before this.
___
*[pkg/storage/replica.go, line 5146 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQ-LliN143I2YCp0yzR:-LQBLodR9A0y-pe41Av4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff5e60d70dfd3422dc474dddcbae9abf17f853c7/pkg/storage/replica.go#L5146)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Definitely WIP. Thanks for pointing it out.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

benesch commented on pkg/storage/replica.go:
```diff
@@ -5148,6 +5148,51 @@ func (r *Replica) sendRaftMessage(ctx context.Context, msg raftpb.Message) {
+		ticks := r.mu.ticks
+		r.mu.RUnlock()
+
+		if msg.RejectHint == 0 && ticks < r.store.cfg.RaftPostSplitSuppressSnapshotTicks {
```
Don’t you want to check at least one of those aforementioned conditions besides ticks? Because ticks resets on node reboot, no?

benesch replied:
I’m on mobile so I can’t actually check this right now.

tschottdorf replied:
In checking that `RejectHint == 0` the replica basically tells us that it has empty state, so it implies `!IsInitialized()`. Or am I misunderstanding you?
I think that's what @bdarnell meant when he said

> Just one of the first three checks should be sufficient (in combination with ticks < 10).

benesch replied:
Oh, I see, I totally misunderstood what RejectHint was. I thought it was some internal Raft thing; didn't realize it was the last Raft index.

I do think `RejectHint == 0` is the least obvious condition to check (`!r.IsInitialized() && r.mu.ticks < 10` is just clearer to me), but I'm also relatively unfamiliar with Raft so if this reads better to you, leave it!

bdarnell reviewed (approved):
> Reviewed 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31875)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/base/config.go, line 116 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQFKkF50iowuF60f04_:-LQFKkF50iowuF60f04a:b28ojhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/32ea73b75388bb643ecaa2aff3f7ea2f4bf46f26/pkg/base/config.go#L116)):*
> ```Go
> 
> 	defaultRaftPostSplitSuppressSnapshotTicks = envutil.EnvOrDefaultInt(
> 		"COCKROACH_RAFT_POST_SPLIT_SUPPRESS_SNAPSHOT_TICKS", 20)
> ```

Does this need an env var? It doesn't seem very environment-dependent and I don't think we should give it an env var just because the neighboring variables have one (I think we should also consider removing the env vars from a lot of these, and changing the ones that are worth keeping to cluster settings except where there are bootstrapping concerns). 
___
*[pkg/storage/replica.go, line 5157 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LPfjWP-pE1_uMZBKfR3-r6-5157:-LQFK86-CIkAnaNT74XP:bzhmoh3) ([raw file](https://github.com/cockroachdb/cockroach/blob/32ea73b75388bb643ecaa2aff3f7ea2f4bf46f26/pkg/storage/replica.go#L5157)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Oh, I see, I totally misunderstood what RejectHint was. I thought it was some internal Raft thing; didn't realize it was the last Raft index.

I do think `RejectHint == 0` is the least obvious condition to check (`!r.IsInitialized() && r.mu.ticks < 10` is just clearer to me), but I'm also relatively unfamiliar with Raft so if this reads better to you, leave it!
</blockquote></details>

`RejectHint==0` is subtle. In some sense it's the closest thing semantically to the check we want, but it's also the default for an uninitialized field. Raft will currently never send a MsgAppResp with Reject but no RejectHint, but that's not a guarantee I feel comfortable relying on. I'd rather use `IsInitialized()`. 


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Still have to write some tests.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-:-LQLuKMT2FT3KKWWOjjW:b-502gnm)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/base/config.go, line 116 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQFKkF50iowuF60f04_:-LQLu-5I57rNrS0OziSK:b8cxfna) ([raw file](https://github.com/cockroachdb/cockroach/blob/32ea73b75388bb643ecaa2aff3f7ea2f4bf46f26/pkg/base/config.go#L116)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Does this need an env var? It doesn't seem very environment-dependent and I don't think we should give it an env var just because the neighboring variables have one (I think we should also consider removing the env vars from a lot of these, and changing the ones that are worth keeping to cluster settings except where there are bootstrapping concerns). 
</blockquote></details>

I don't know whether it needs one. We have one for the tick interval, and the mechanism here depends on that. For example, if someone changed the Raft tick interval to 1s, would they want to stall uninitialized replicas for 20s? Probably not. I generally agree that nobody uses or should use these knobs, but for the near future it might be useful for my testing, so I'll leave it unless you feel strongly.
___
*[pkg/storage/replica.go, line 5157 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LPfjWP-pE1_uMZBKfR3-r6-5157:-LQLtudD-co_WHqGOc3Q:bq7gyfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/32ea73b75388bb643ecaa2aff3f7ea2f4bf46f26/pkg/storage/replica.go#L5157)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`RejectHint==0` is subtle. In some sense it's the closest thing semantically to the check we want, but it's also the default for an uninitialized field. Raft will currently never send a MsgAppResp with Reject but no RejectHint, but that's not a guarantee I feel comfortable relying on. I'd rather use `IsInitialized()`. 
</blockquote></details>

Done. I agree that it's subtle, especially since Raft internally uses the hint only if it's nonzero.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31875)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/base/config.go, line 116 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LQFKkF50iowuF60f04_:-LQop-wZA8WnwXyZTpp2:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/32ea73b75388bb643ecaa2aff3f7ea2f4bf46f26/pkg/base/config.go#L116)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I don't know whether it needs one. We have one for the tick interval, and the mechanism here depends on that. For example, if someone changed the Raft tick interval to 1s, would they want to stall uninitialized replicas for 20s? Probably not. I generally agree that nobody uses or should use these knobs, but for the near future it might be useful for my testing, so I'll leave it unless you feel strongly.
</blockquote></details>

OK


<!-- Sent from Reviewable.io -->

tbg committed:
storage: never truncate log for inflight snapshots

When a (preemptive or Raft) snapshot is inflight, a range should not
truncate its log index to one ahead of the pending snapshot as doing so
would require yet another snapshot.

The code prior to this commit attempted to achieve that by treating the
pending snapshot as an additional follower in the computation of the
quorum index. This works as intended as long as the Raft log is below
its configured maximum size but usually leads to abandoning the snapshot
index when truncating based on size (to the quorum commit index). This
situation occurs frequently during the split+scatter phase of data
imports, where (mostly empty) ranges are rapidly upreplicated and split.

This isn't limited to small replicas, however. Assume a range is ~32mb
in size and a (say, preemptive) snapshot is underway. Preemptive
snapshots are (at the time of writing) throttled to 2mb/s, so it will
take approximately 16s to go through. At the same time, the Raft log may
easily grow in size by the size truncation threshold (4mb at time of
writing), allowing a log truncation that abandons the snapshot. A
similar phenomenon applies to Raft snapshots, though now the quota pool
will place restrictions on forward Raft progress, however not if a
single command exceeds the size restriction (as is common during RESTORE
operations where individual Raft commands are large). I haven't
conclusively observed this in practice (though there has been enough
badness to suspect that it happened), but in principle this could lead
to a potentially infinite number of snapshots being sent out, a very
expensive way of keeping a follower up to date.

After this change, the pending snapshot index is managed more carefully
(and is managed for Raft snapshots as well, not only for preemptive
ones) and is never truncated away. As a result, in regular operation
snapshots should now be able to be followed by regular Raft log catch-up
in all cases. More precisely, we keep a map of pending snapshot index to
deadline. The deadline is zero while the snapshot is ongoing and is set
when the snapshot is completed. This avoids races between the snapshot
completing and the replication change completing (which would occur
even if the snapshot is only registering as completed after the
replication change completes).

There's an opportunity for a bigger refactor here by using learner
replicas instead of preemptive snapshots. The idea is to add the replica
as a learner first (ignoring it for the quota pool until it has received
the first snapshot) first, and to use the regular Raft snapshot
mechanism to catch it up. Once achieved, another configuration change
would convert the learner into a regular follower. This change in itself
will likely make our code more principled, but it is a more invasive
change that is left for the future.

Similarly, there is knowledge in the quota pool that it seems we should
be using for log truncation decisions. The quota pool essentially knows
about the size of each log entry whereas the Raft truncations only know
the accumulated approximate size of the full log. For instance, instead
of blindly truncating to the quorum commit index when the Raft log is
too large, we could truncate it to an index that reduces the log size to
about 50% of the maximum, in effect reducing the number of snapshots
that are necessary due to quorum truncation. It's unclear whether this
improvement will matter in practice.

The following script reproduces redundant Raft snapshots (needs a
somewhat beefy machine such as a gceworker). Note that this script will
also incur Raft snapshots due to the splits, which are fixed in a
follow-up commit.

```
set -euxo pipefail

killall -9 cockroach || true
killall -9 workload || true
sleep 1
rm -rf cockroach-data || true
mkdir -p cockroach-data

./cockroach start --insecure --host=localhost --port=26257 --http-port=26258 --store=cockroach-data/1 --cache=256MiB --background
./cockroach start --insecure --host=localhost --port=26259 --http-port=26260 --store=cockroach-data/2 --cache=256MiB --join=localhost:26257 --background
./cockroach start --insecure --host=localhost --port=26261 --http-port=26262 --store=cockroach-data/3 --cache=256MiB --join=localhost:26257 --background
./cockroach start --insecure --host=localhost --port=26263 --http-port=26264 --store=cockroach-data/4 --cache=256MiB --join=localhost:26257 --background

sleep 5

./cockroach sql --insecure -e 'set cluster setting kv.range_merge.queue_enabled = false;'

./cockroach sql --insecure <<EOF
CREATE DATABASE csv;
IMPORT TABLE csv.lineitem CREATE USING 'gs://cockroach-fixtures/tpch-csv/schema/lineitem.sql'
CSV DATA (
 'gs://cockroach-fixtures/tpch-csv/sf-1/lineitem.tbl.1',
 'gs://cockroach-fixtures/tpch-csv/sf-1/lineitem.tbl.2',
 'gs://cockroach-fixtures/tpch-csv/sf-1/lineitem.tbl.3',
 'gs://cockroach-fixtures/tpch-csv/sf-1/lineitem.tbl.4',
 'gs://cockroach-fixtures/tpch-csv/sf-1/lineitem.tbl.5',
 'gs://cockroach-fixtures/tpch-csv/sf-1/lineitem.tbl.6',
 'gs://cockroach-fixtures/tpch-csv/sf-1/lineitem.tbl.7',
 'gs://cockroach-fixtures/tpch-csv/sf-1/lineitem.tbl.8'
) WITH delimiter='|' ;
EOF

sleep 5

for port in 26257 26259 26261 26263; do
 ./cockroach sql --insecure -e "select name, value from crdb_internal.node_metrics where name like '%raftsn%' order by name desc" --port "${port}"
+done
```

Release note: None

tschottdorf commented:
Ready for another look. I'll need to do another pass for cleanliness (factor out constants, add some comments, unit test the deadline expiration) but this is what I'm going to be testing the nightlies with.

I added a high-level test that runs 100 splits on a 3 node cluster (passes in 1-3s usually) and verifies that there aren't any Raft snapshots. Due to catching the initial upreplication, it also tests (though less decisively) that Raft log truncation isn't overly aggressive. To get the test stable under `stress`, I've had to add another commit that respects snapshots more strongly during log truncations.

That other commit might as well be the decisive one. I'm going to run another suite of known-flaky import tests with both.

I also haven't observed https://github.com/cockroachdb/cockroach/pull/32139 play a noticeable role here. Even though that remains on my radar, I don't think it needs to be fixed before this merges.

tschottdorf commented:
@nvanbenschoten we chatted about the log truncation size heuristics being overly aggressive last week and I agree with that. With the newly introduced logging (just merged to master), I see these quite frequently:

```
I181112 16:23:26.968872 2407 storage/raft_log_queue.go:367  [n1,raftlog,s1,r4/1:/System/{NodeLive…-tsd}] truncate 6 entries to first index 108 (chosen via: quorum); log too large (89 KiB > 72 KiB); implies 1 Raft snapshot
```

I believe these don't actually usually lead to snapshots because the follower beats the log truncation deterministically (when there aren't snapshots etc) but it sure would be nice to not cut off the follower like that (I think the quota pool can help us here).

There's also the question of what impact on performance it has to truncate less/more aggressively. I assume a write-heavy three node kv test would be the right testbed for this, right?

I'm not planning on changing anything else in this PR because it's already changing a lot, but I do want to follow on up these heuristics.

tschottdorf commented:
Hmm, I did get another Raft snapshot with a [small script](https://gist.github.com/tschottdorf/0580cc7a7fa29b68a4200dcdf54e327e) I like to run to shake out problems. Happened soon after a rebalance, and I'm not seeing a log truncation nor the message about letting an MsgAppResp rejection through.

https://gist.github.com/tschottdorf/293bde17ddc05dacdd476ef4dea9956e

benesch commented:
Fixed link to small script: https://gist.github.com/tbg/0580cc7a7fa29b68a4200dcdf54e327e

tbg commented:
Small script was the wrong small script. Updated the gist.

nvanbenschoten commented:
> I believe these don't actually usually lead to snapshots because the follower beats the log truncation deterministically

Is this before or after we send the follower an entire snapshot?

> I assume a write-heavy three node kv test would be the right testbed for this, right?

Yes, both with small and large writes.

benesch reviewed (commented):
> Reviewed 6 of 6 files at r7, 3 of 3 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31875)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 361 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8__my9qZ4eo451_7V:-LR8__my9qZ4eo451_7W:b-jk6o18) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f6dd0780063e9a24e220b2ca88ad600dede5a4b/pkg/storage/replica.go#L361)):*
> ```Go
> 		// already finished snapshot "pending" for extended periods of time
> 		// (preventing log truncation).
> 		pendingSnapshotIndex map[uint64]time.Time
> ```

Can we rename this to `pendingSnapshotIndexes` or similar?
___
*[pkg/storage/replica.go, line 361 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8cZnY06RnZi1fnOL3:-LR8cZnY06RnZi1fnOL4:bq1a7qq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f6dd0780063e9a24e220b2ca88ad600dede5a4b/pkg/storage/replica.go#L361)):*
> ```Go
> 		// already finished snapshot "pending" for extended periods of time
> 		// (preventing log truncation).
> 		pendingSnapshotIndex map[uint64]time.Time
> ```

Keying this by pending index is kinda weird. Why not map snapshot ID to {deadline, index}? Then you won't need to comment about why it's mostly OK that you might have two snapshots that clobber each other.
___
*[pkg/storage/replica.go, line 6954 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8bVIW2i0XEJ6KmnOp:-LR8bVIW2i0XEJ6KmnOq:bp7ycvc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f6dd0780063e9a24e220b2ca88ad600dede5a4b/pkg/storage/replica.go#L6954)):*
> ```Go
> }
> 
> func (r *Replica) setPendingSnapshotIndexLocked(index uint64) {
> ```

Consider renaming to `addPendingSnapshot` or similar.
___
*[pkg/storage/replica.go, line 6971 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8a8wP5GC_KORYjb98:-LR8a8wP5GC_KORYjb99:bhyciie) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f6dd0780063e9a24e220b2ca88ad600dede5a4b/pkg/storage/replica.go#L6971)):*
> ```Go
> 	if r.mu.pendingSnapshotIndex == nil {
> 		r.mu.pendingSnapshotIndex = make(map[uint64]time.Time)
> 	}
> ```

nit: I find it somewhat weird that you read from the map before initializing it. (I know reading from nil maps is safe even though writing to them is not—thanks, Go—but I still find it weird.) Don't bother changing it unless you also find it weird now that I've pointed it out.
___
*[pkg/storage/replica.go, line 6976 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8bpno-CudzTvhZI3W:-LR8bpno-CudzTvhZI3X:bqsfppt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f6dd0780063e9a24e220b2ca88ad600dede5a4b/pkg/storage/replica.go#L6976)):*
> ```Go
> }
> 
> func (r *Replica) clearPendingSnapshotIndex(origIndex uint64) {
> ```

Consider renaming to "markPendingSnapshotCompleted."
___
*[pkg/storage/replica.go, line 6980 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8bfi95KR_hjhOOQaS:-LR8bfi95KR_hjhOOQaT:b-62ferj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f6dd0780063e9a24e220b2ca88ad600dede5a4b/pkg/storage/replica.go#L6980)):*
> ```Go
> 	defer r.mu.Unlock()
> 
> 	deadline := timeutil.Now().Add(3 * time.Second)
> ```

Perhaps explain why 3s.
___
*[pkg/storage/replica.go, line 5178 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8eBUp7enofaXPShdi:-LR8eBUp7enofaXPShdj:bwpnnsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b46b9df7a3dc9f7254bef74116df6c06e7a925b/pkg/storage/replica.go#L5178)):*
> ```Go
> 	// 1. a rebalance operation is adding a new replica of the range to this
> 	// node. We always send a preemptive snapshot before attempting to do so, so
> 	// we wouldn't enter this branch as desc.Replicas would be nonempty. We
> ```

s/desc.Replicas would be nonempty/initialized would be true/
___
*[pkg/storage/replica_raftstorage.go, line 402 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8d-cN2qtT1lG5JpEf:-LR8d-cN2qtT1lG5JpEg:bv8pno8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f6dd0780063e9a24e220b2ca88ad600dede5a4b/pkg/storage/replica_raftstorage.go#L402)):*
> ```Go
> 	r.mu.Lock()
> 	appliedIndex := r.mu.state.RaftAppliedIndex
> 	r.setPendingSnapshotIndexLocked(appliedIndex) // cleared when OutgoingSnapshot closes
> ```

As mentioned above, I think keying pending snapshots by snap UUID would be cleaner. Unless I'm missing something, it should be straightforward to hoist the UUID generation out of `snapshot()`.


<!-- Sent from Reviewable.io -->

tbg commented:
No updates to the code yet, but I did take this for a spin:

```
--- PASS: import/tpch/nodes=4 (16690.89s)
```

Going to run a few more of those overnight.

tbg commented:
Got another four passes overnight:

--- PASS: import/tpch/nodes=4 (17664.83s)
--- PASS: import/tpch/nodes=4 (18798.21s)
--- PASS: import/tpch/nodes=4 (18108.33s)
--- PASS: import/tpch/nodes=4 (18004.14s)

That's pretty encouraging. Going to kick of four runs that don't have this PR (but are otherwise identical).

tbg commented:
Note the two pending commands here. One follower obviously needs a snapshot. The other follower does too, even though it shouldn't actually need one (index 10 is the truncated index, which means that the leader should have index 11 and could send it along). Not clear how this came to be, but one replica was removed from this range. Perhaps the leader plus that now removed replica made progress, so the group wedged itself when it downreplicated off of that functioning follower. Both of these problems are already mentioned in https://github.com/cockroachdb/cockroach/issues/32046#issue-375725955.

![image](https://user-images.githubusercontent.com/5076964/48482376-d1190a00-e810-11e8-96bd-f36df4d1a168.png)

tbg commented:
I'm also noticing that the spikes in snapshots are per node (i.e. the cluster view is a superposition of per-node spikes). This lends further credence to this being caused by splitting ranges that already need a snapshot.

tbg reviewed (commented):
Updated, PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-:-LRHMH2PDbNJ7iwt97QR:b-rr168b)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 361 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8__my9qZ4eo451_7V:-LRGwrJ243WB7C7G_bpw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f6dd0780063e9a24e220b2ca88ad600dede5a4b/pkg/storage/replica.go#L361)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Can we rename this to `pendingSnapshotIndexes` or similar?
</blockquote></details>

Done.
___
*[pkg/storage/replica.go, line 361 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8cZnY06RnZi1fnOL3:-LRGwrun-Mfsjlzh4iM9:b-4ivtvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f6dd0780063e9a24e220b2ca88ad600dede5a4b/pkg/storage/replica.go#L361)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Keying this by pending index is kinda weird. Why not map snapshot ID to {deadline, index}? Then you won't need to comment about why it's mostly OK that you might have two snapshots that clobber each other.
</blockquote></details>

You're right, this was silly all along. Done.
___
*[pkg/storage/replica.go, line 6971 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8a8wP5GC_KORYjb98:-LRGmud8CZfd-iaPcC9t:bnuo0xr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f6dd0780063e9a24e220b2ca88ad600dede5a4b/pkg/storage/replica.go#L6971)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

nit: I find it somewhat weird that you read from the map before initializing it. (I know reading from nil maps is safe even though writing to them is not—thanks, Go—but I still find it weird.) Don't bother changing it unless you also find it weird now that I've pointed it out.
</blockquote></details>

I did that intentionally, but there wasn't a real good reason. Done.
___
*[pkg/storage/replica.go, line 6976 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8bpno-CudzTvhZI3W:-LRGn1OF7WG7EihYErGL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f6dd0780063e9a24e220b2ca88ad600dede5a4b/pkg/storage/replica.go#L6976)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Consider renaming to "markPendingSnapshotCompleted."
</blockquote></details>

Done.
___
*[pkg/storage/replica.go, line 6980 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8bfi95KR_hjhOOQaS:-LRGnlMn1czt2cYrqQ8W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f6dd0780063e9a24e220b2ca88ad600dede5a4b/pkg/storage/replica.go#L6980)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Perhaps explain why 3s.
</blockquote></details>

Done.
___
*[pkg/storage/replica.go, line 5178 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8eBUp7enofaXPShdi:-LRGm9H98rbn-Rys8L5E:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b46b9df7a3dc9f7254bef74116df6c06e7a925b/pkg/storage/replica.go#L5178)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

s/desc.Replicas would be nonempty/initialized would be true/
</blockquote></details>

Done.
___
*[pkg/storage/replica_raftstorage.go, line 402 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-LR8d-cN2qtT1lG5JpEf:-LRGwv5l0T55SQx_LHxn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f6dd0780063e9a24e220b2ca88ad600dede5a4b/pkg/storage/replica_raftstorage.go#L402)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

As mentioned above, I think keying pending snapshots by snap UUID would be cleaner. Unless I'm missing something, it should be straightforward to hoist the UUID generation out of `snapshot()`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg commented:
An additional four instances of `import/tpch/nodes=4` just finished successfully. 🎆

--- PASS: import/tpch/nodes=4 (17403.51s)
--- PASS: import/tpch/nodes=4 (18198.41s)
--- PASS: import/tpch/nodes=4 (19152.02s)
--- PASS: import/tpch/nodes=4 (18090.76s)

benesch reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 6 of 6 files at r9, 3 of 3 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31875#-:-LRIakjG4VW60kXsJKFf:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

tbg committed:
storage: avoid errant Raft snapshots after splits

A known race occurs during splits when some nodes apply the split
trigger faster than others. The "slow" node(s) may learn about the
newly created right hand side replica through Raft messages arriving
from the "fast" nodes. In such cases, the leader will immediately try to
catch up the follower (which it sees at log position zero) via a
snapshot, but this isn't possible since there's an overlapping replica
(the pre-split replica waiting to apply the trigger). This both leads to
unnecessary transfer of data and can clog the Raft snapshot queue which
tends to get stuck due to the throttling mechanisms both at the sender
and receivers.

To prevent this race (or make it exceedingly unlikely), we selectively
drop certain messages from uninitialized followers, namely those that
refuse an append to the log, for a number of ticks (corresponding to
at most a few seconds of real time). Not dropping such a message leads
to a Raft snapshot as the leader will learn that the follower has last
index zero, which is never an index that can be caught up to from the
log (our log "starts" at index 10).

The script below reproduces the race (prior to this commit) by running
1000 splits back to back in a three node local cluster, usually showing
north of a hundred Raft snapshots, i.e. a >10% chance to hit the race
for each split. There's also a unit test that exposes this problem and
can be stressed more conveniently (it also exposes the problems in the
preceding commit related to overly aggressive log truncation).

The false positives here are a) the LHS of the split needs a snapshot
which catches it up across the split trigger and b) the LHS is
rebalanced away (and GC'ed) before applying the split trigger. In both
cases the timeout-based mechanism would allow the snapshot after a few
seconds, once the Raft leader contacts the follower for the next time.

Note that the interaction with Raft group quiescence is benign. We're
only dropping MsgAppResp which is only sent by followers, implying that
the Raft group is already unquiesced.

```
set -euxo pipefail

killall -9 cockroach || true
killall -9 workload || true
sleep 1
rm -rf cockroach-data || true
mkdir -p cockroach-data

./cockroach start --insecure --host=localhost --port=26257 --http-port=26258 --store=cockroach-data/1 --cache=256MiB --background
./cockroach start --insecure --host=localhost --port=26259 --http-port=26260 --store=cockroach-data/2 --cache=256MiB --join=localhost:26257 --background
./cockroach start --insecure --host=localhost --port=26261 --http-port=26262 --store=cockroach-data/3 --cache=256MiB --join=localhost:26257 --background

sleep 5

./cockroach sql --insecure -e 'set cluster setting kv.range_merge.queue_enabled = false;'
./bin/workload run kv --splits 1000 --init --drop --max-ops 1

sleep 5

for port in 26257 26259 26261; do
  ./cockroach sql --insecure -e "select name, value from crdb_internal.node_metrics where name like '%raftsn%' order by name desc" --port "${port}"
done
```

Release note (bug fix): Avoid occasional unnecessary Raft snapshots after Range splits.

ajwerner commented:
bors r=benesch
